File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/elements/InputPercent.java
package info.nightscout.androidaps.plugins.general.automation.elements;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.LinearLayout;

import java.text.DecimalFormat;

import info.nightscout.androidaps.utils.NumberPicker;

public class InputPercent extends Element {

    final TextWatcher textWatcher = new TextWatcher() {
        @Override
        public void afterTextChanged(Editable s) {
            value = Math.max(value, 70d);
            value = Math.min(value, 130d);
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
        }
    };

    private double value;

    public InputPercent() {
        super();
        value = 100d;
    }

    public InputPercent(InputPercent another) {
        super();
        value = another.getValue();
    }

    @Override
    public void addToLayout(LinearLayout root) {
        NumberPicker numberPicker = new NumberPicker(root.getContext(), null);
        numberPicker.setParams(100d, 70d, 130d, 5d, new DecimalFormat("0"), true, null, textWatcher);
        numberPicker.setValue(value);
        numberPicker.setOnValueChangedListener(value -> this.value = value);
        root.addView(numberPicker);
    }

    public double getValue() {
        return value;
    }

    public InputPercent setValue(double value) {
        this.value = value;
        return this;
    }
}

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/elements/InputPercent.kt
package info.nightscout.androidaps.plugins.general.automation.elements

import android.widget.LinearLayout
import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.R
import info.nightscout.androidaps.utils.ui.NumberPicker
import java.text.DecimalFormat

class InputPercent(injector: HasAndroidInjector) : Element(injector) {
    var value: Double = 100.0

    constructor(injector: HasAndroidInjector, value: Double) : this(injector) {
        this.value = value
    }

    override fun addToLayout(root: LinearLayout) {
        val numberPicker = NumberPicker(root.context, null)
        numberPicker.setParams(100.0, MIN, MAX, 5.0, DecimalFormat("0"), true, root.findViewById(R.id.ok))
        numberPicker.value = value
        numberPicker.setOnValueChangedListener { value: Double -> this.value = value }
        root.addView(numberPicker)
    }

    companion object {
        const val MIN = 70.0
        const val MAX = 130.0
    }
}